server/src/auth/auth.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable } from '@nestjs/common';
+import { JwtService } from '@nestjs/jwt';
+import fetch from 'node-fetch';
+import * as rs from 'randomstring';
+import { AppConfigService } from 'src/config/app.service';
+import { AuthConfigService } from 'src/config/auth-config.service';
+import { User } from 'src/models/users/entities/user.entity';
+import { UserService } from 'src/models/users/user.service';
+
+const STATE = rs.generate(20);
+
+const KAKAO_AUTH_URI = 'https://kauth.kakao.com/oauth/authorize';
+const KAKAO_TOKEN_URI = 'https://kauth.kakao.com/oauth/token';
+const KAKAO_API_URI = 'https://kapi.kakao.com/v2/user/me';
+const KAKAO_CONTENT_TYPE = 'application/x-www-form-urlencoded';
+
+@Injectable()
+export class AuthService {
+  private REDIRECT_URI: string;
+  private REST_API_KEY: string;
+
+  constructor(
+    private jwtService: JwtService,
+    private userService: UserService,
+    authConfigService: AuthConfigService,
+    appConfigService: AppConfigService,
+  ) {
+    this.REST_API_KEY = authConfigService.restApi;
+    this.REDIRECT_URI = appConfigService.server + '/api/v1/auth/redirect';
+  }
+
+  getRequestUrl() {
+    const querystring = `?client_id=${this.REST_API_KEY}&redirect_uri=${this.REDIRECT_URI}&state=${STATE}&response_type=code`;
+    return `${KAKAO_AUTH_URI}${querystring}`;
+  }
+
+  async oauth(code: string, state: string) {
+    if (state !== STATE) throw new Error('STATE가 다릅니다. csrf 공격이 의심됩니다.');
+
+    const param = new URLSearchParams({
+      grant_type: 'authorization_code',
+      client_id: this.REST_API_KEY,
+      redirect_uri: this.REDIRECT_URI,
+      code,
+    });
+
+    const response = await fetch(KAKAO_TOKEN_URI, {
+      method: 'post',
+      body: param,
+      headers: {
+        'Content-Type': KAKAO_CONTENT_TYPE,
+      },
+    });
+
+    const result = await response.json();
+    return this.getUserInfo(result.access_token);
+  }
+
+  async getUserInfo(accessToken: string) {
+    const response = await fetch(KAKAO_API_URI, {
+      headers: {
+        Accept: `${KAKAO_CONTENT_TYPE};charset=utf-8`,
+        Authorization: `Bearer ${accessToken}`,
+      },
+    });
+
+    const result = await response.json();
+
+    const oAuthId: string = result.id;
+    const email: string = result.kakao_account.email;
+    return { accessToken, oAuthId, email };
+  }
+
+  async validateUser(loginId: string, password: string): Promise<Partial<User>> {
```

Comment:
```suggestion
  async validateUser(loginId: string, password: string): Promise<Partial<User> | null> {
```

반환타입에 `null`이 빠지면 반드시 호출시 `Partial<User>`를 넘겨주거나 그렇지 못하면 throw를 해야합니다.
대부분의 라이브러리나 기본 js도 그렇게 돼 있어요.

Reply:
null 도 Partial<User>의 일부라고 생각했는데, 말씀대로 변경했습니다!